File: GameLayers/gameScorer.js
import {config} from '../config.js'

export class DefaultScorer {
    constructor(healthPower){
        this.healthPower = healthPower
    }
    scoreGame = function(isVictory, world) {
        var baseScore = config.SCORE_MULTIPLIER
        var score = baseScore * (isVictory ? Math.pow(world.hero.health, this.healthPower) 
            : -Math.pow(world.boss.health, Math.floor(this.healthPower)))
        return score
    }
}

export class ExpScorer {
    constructor(powerConstant, linearConstant){    
        this.powerConstant = powerConstant
        this.linearConstant = linearConstant
    }

    scoreGame = function(isVictory, world) {
        var hero = Math.max(0, world.hero.health)
        var boss = Math.max(0, world.boss.health)
        var score = (isVictory ? 1 : -1)*Math.pow(10, this.powerConstant*(hero+boss))
        return score*this.linearConstant;
    }
}

export class LogScorer {
    constructor(c1, c2){
        this.c1 = c1
        this.c2 = c2
    }

    scoreGame = function(isVictory, world) {
        var hero = Math.max(0, world.hero.health)
        var boss = Math.max(0, world.boss.health)
        var score = isVictory?1:-1
        score *= this.c1*(Math.log(hero+boss))
        return score
    }
}

File: python/GameLayers/gameScorer.py
import math
import config

class DefaultScorer:
    def __init__(self, healthPower):
        self.healthPower = healthPower

    def scoreGame(self, isVictory, world):
        baseScore = config.SCORE_MULTIPLIER
        victoryScore = math.pow(world.hero.health, self.healthPower)
        defeatScore = -math.pow(world.boss.health, math.floor(self.healthPower))
        score = baseScore * (victoryScore if isVictory else defeatScore)
        return score

class ExpScorer:
    def __init__(self, powerConstant, linearConstant):    
        self.powerConstant = powerConstant
        self.linearConstant = linearConstant

    def scoreGame(self, isVictory, world):
        hero = max(0, world.hero.health)
        boss = max(0, world.boss.health)
        score = (1 if isVictory else -1)*math.pow(2, self.powerConstant*(hero+boss))
        return score*self.linearConstant;

class LogScorer:
    def __init__(self, c1, c2):
        self.c1 = c1
        self.c2 = c2

    def scoreGame(self, isVictory, world):
        hero = max(0, world.hero.health)
        boss = max(0, world.boss.health)
        score = 1 if isVictory else -1
        score *= self.c1*(math.log(hero+boss))
        return score

